File: python/features/steps/discovery_steps.py
"""Behave steps for project discovery scenarios."""

from __future__ import annotations

from pathlib import Path

from behave import given, then

from features.steps.shared import build_issue, ensure_git_repository, write_issue_file
from taskulus.ids import format_issue_key


def _create_repo(context: object, name: str) -> Path:
    root = Path(context.temp_dir) / name
    root.mkdir(parents=True, exist_ok=True)
    ensure_git_repository(root)
    context.working_directory = root
    return root


def _write_issue(project_dir: Path, identifier: str, title: str) -> None:
    issues_dir = project_dir / "issues"
    issues_dir.mkdir(parents=True, exist_ok=True)
    issue = build_issue(identifier, title, "task", "open", None, [])
    write_issue_file(project_dir, issue)


@given("a repository with nested project directories")
def given_repo_with_nested_projects(context: object) -> None:
    root = _create_repo(context, "nested-projects")
    root_project = root / "project"
    nested_project = root / "nested" / "project"
    _write_issue(root_project, "tsk-root", "Root task")
    _write_issue(nested_project, "tsk-nested", "Nested task")
    context.discovered_issue_keys = [
        format_issue_key("tsk-root", project_context=False),
        format_issue_key("tsk-nested", project_context=False),
    ]


@given("a repository with a project directory above the current directory")
def given_repo_project_above_cwd(context: object) -> None:
    root = _create_repo(context, "project-above")
    project_dir = root / "project"
    _write_issue(project_dir, "tsk-above", "Above task")
    child_dir = root / "child"
    child_dir.mkdir(parents=True, exist_ok=True)
    context.working_directory = child_dir
    context.above_issue_key = format_issue_key("tsk-above", project_context=True)


@given("a project directory with a sibling project-local directory")
def given_repo_with_project_local(context: object) -> None:
    root = _create_repo(context, "project-local")
    project_dir = root / "project"
    local_dir = root / "project-local"
    _write_issue(project_dir, "tsk-shared1", "Shared task")
    _write_issue(local_dir, "tsk-local1", "Local task")
    context.shared_issue_key = format_issue_key("tsk-shared1", project_context=True)
    context.local_issue_key = format_issue_key("tsk-local1", project_context=True)


@given("a repository with a .taskulus file referencing another project")
def given_repo_with_taskulus_dotfile(context: object) -> None:
    root = _create_repo(context, "taskulus-dotfile")
    external_root = Path(context.temp_dir) / "dotfile-external"
    external_project = external_root / "project"
    _write_issue(external_project, "tsk-external", "External task")
    (root / ".taskulus").write_text(
        f"{external_project}\n",
        encoding="utf-8",
    )
    context.external_project_path = external_project.resolve()


@given("a repository with a .taskulus file referencing a missing path")
def given_repo_with_taskulus_dotfile_missing(context: object) -> None:
    root = _create_repo(context, "taskulus-dotfile-missing")
    (root / ".taskulus").write_text("missing/project\n", encoding="utf-8")


@then("issues from all discovered projects should be listed")
def then_issues_from_all_projects(context: object) -> None:
    for key in getattr(context, "discovered_issue_keys", []):
        assert key in context.result.stdout


@then("no issues should be listed")
def then_no_issues_listed(context: object) -> None:
    above_key = getattr(context, "above_issue_key", None)
    if above_key:
        assert above_key not in context.result.stdout
        return
    assert context.result.stdout.strip() == ""


@then("local issues should be included")
def then_local_issues_included(context: object) -> None:
    assert context.local_issue_key in context.result.stdout


@then("local issues should not be listed")
def then_local_issues_not_listed(context: object) -> None:
    assert context.local_issue_key not in context.result.stdout


@then("only local issues should be listed")
def then_only_local_issues_listed(context: object) -> None:
    assert context.local_issue_key in context.result.stdout
    assert context.shared_issue_key not in context.result.stdout


@then("issues from the referenced project should be listed")
def then_issues_from_referenced_project_listed(context: object) -> None:
    expected = format_issue_key("tsk-external", project_context=False)
    assert expected in context.result.stdout

File: rust/features/steps/discovery_steps.rs
use std::fs;
use std::path::PathBuf;
use std::process::Command;

use chrono::{TimeZone, Utc};
use cucumber::{given, then};

use taskulus::ids::format_issue_key;
use taskulus::models::IssueData;

use crate::step_definitions::initialization_steps::TaskulusWorld;

fn create_repo(world: &mut TaskulusWorld, name: &str) -> PathBuf {
    let temp_dir = tempfile::TempDir::new().expect("tempdir");
    let repo_path = temp_dir.path().join(name);
    fs::create_dir_all(&repo_path).expect("create repo dir");
    Command::new("git")
        .args(["init"])
        .current_dir(&repo_path)
        .output()
        .expect("git init failed");
    world.working_directory = Some(repo_path.clone());
    world.temp_dir = Some(temp_dir);
    repo_path
}

fn build_issue(identifier: &str, title: &str) -> IssueData {
    let timestamp = Utc.with_ymd_and_hms(2026, 2, 11, 0, 0, 0).unwrap();
    IssueData {
        identifier: identifier.to_string(),
        title: title.to_string(),
        description: String::new(),
        issue_type: "task".to_string(),
        status: "open".to_string(),
        priority: 2,
        assignee: None,
        creator: None,
        parent: None,
        labels: Vec::new(),
        dependencies: Vec::new(),
        comments: Vec::new(),
        created_at: timestamp,
        updated_at: timestamp,
        closed_at: None,
        custom: std::collections::BTreeMap::new(),
    }
}

fn write_issue(project_dir: &PathBuf, issue: &IssueData) {
    let issues_dir = project_dir.join("issues");
    fs::create_dir_all(&issues_dir).expect("create issues dir");
    let issue_path = issues_dir.join(format!("{}.json", issue.identifier));
    let contents = serde_json::to_string_pretty(issue).expect("serialize issue");
    fs::write(issue_path, contents).expect("write issue");
}

#[given("a repository with nested project directories")]
fn given_repo_nested_projects(world: &mut TaskulusWorld) {
    let root = create_repo(world, "nested-projects");
    let root_project = root.join("project");
    let nested_project = root.join("nested").join("project");
    write_issue(&root_project, &build_issue("tsk-root", "Root task"));
    write_issue(&nested_project, &build_issue("tsk-nested", "Nested task"));
}

#[given("a repository with a project directory above the current directory")]
fn given_repo_project_above_cwd(world: &mut TaskulusWorld) {
    let root = create_repo(world, "project-above");
    let project_dir = root.join("project");
    write_issue(&project_dir, &build_issue("tsk-above", "Above task"));
    let child_dir = root.join("child");
    fs::create_dir_all(&child_dir).expect("create child dir");
    world.working_directory = Some(child_dir);
}

#[given("a project directory with a sibling project-local directory")]
fn given_repo_with_project_local(world: &mut TaskulusWorld) {
    let root = create_repo(world, "project-local");
    let project_dir = root.join("project");
    let local_dir = root.join("project-local");
    write_issue(&project_dir, &build_issue("tsk-shared1", "Shared task"));
    write_issue(&local_dir, &build_issue("tsk-local1", "Local task"));
}

#[then("issues from all discovered projects should be listed")]
fn then_issues_from_all_projects_listed(world: &mut TaskulusWorld) {
    let stdout = world.stdout.as_ref().expect("stdout");
    let root_key = format_issue_key("tsk-root", false);
    let nested_key = format_issue_key("tsk-nested", false);
    assert!(stdout.contains(&root_key));
    assert!(stdout.contains(&nested_key));
}

#[then("no issues should be listed")]
fn then_no_issues_listed(world: &mut TaskulusWorld) {
    let stdout = world.stdout.as_ref().expect("stdout");
    assert!(stdout.trim().is_empty());
}

#[then("local issues should be included")]
fn then_local_issues_included(world: &mut TaskulusWorld) {
    let stdout = world.stdout.as_ref().expect("stdout");
    let local_key = format_issue_key("tsk-local1", true);
    assert!(stdout.contains(&local_key));
}

#[then("local issues should not be listed")]
fn then_local_issues_not_listed(world: &mut TaskulusWorld) {
    let stdout = world.stdout.as_ref().expect("stdout");
    let local_key = format_issue_key("tsk-local1", true);
    assert!(!stdout.contains(&local_key));
}

#[then("only local issues should be listed")]
fn then_only_local_issues_listed(world: &mut TaskulusWorld) {
    let stdout = world.stdout.as_ref().expect("stdout");
    let local_key = format_issue_key("tsk-local1", true);
    let shared_key = format_issue_key("tsk-shared1", true);
    assert!(stdout.contains(&local_key));
    assert!(!stdout.contains(&shared_key));
}

#[then("issues from the referenced project should be listed")]
fn then_issues_from_referenced_project(world: &mut TaskulusWorld) {
    let stdout = world.stdout.as_ref().expect("stdout");
    let external_key = format_issue_key("tsk-external", false);
    assert!(stdout.contains(&external_key));
}
